src/main/java/org/commcare/cases/model/StorageIndexedTreeElementModel.java

```diff
@@ -0,0 +1,151 @@
+import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.javarosa.core.util.externalizable.ExtWrapList;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * A DB model for storing TreeElements such that particular attributes and
+ * elements are indexed and queryable using the DB.
+ *
+ * Indexed attributes/elements get their own table columns, and the rest of
+ * the TreeElement is stored as a serialized blob.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class StorageIndexedTreeElementModel implements Persistable, IMetaData {
+
+    private static final String STORAGE_KEY_PREFIX = "IND_FIX_";
+    private static final String DASH_ESCAPE = "\\$\\$";
+    private static final String ATTR_PREFIX = "_$";
+    private static final String ELEM_PREFIX = "_0";
+
+    private String[] metaDataFields = null;
+    private Vector<String> indices;
+    private TreeElement root;
+
+    protected int recordId = -1;
+    protected String entityId;
+
+    @SuppressWarnings("unused")
+    public StorageIndexedTreeElementModel() {
+        // for serialization
+    }
+
+    public StorageIndexedTreeElementModel(Set<String> indices, TreeElement root) {
+        this.indices = new Vector<>(indices);
+        this.root = root;
+        metaDataFields = buildMetadataFields(this.indices);
+    }
+
+    private static String[] buildMetadataFields(List<String> indices) {
+        String[] escapedIndexList = new String[indices.size()];
+        int i = 0;
+        for (String index : indices) {
+            escapedIndexList[i++] = getColFromEntry(index);
+        }
+        return escapedIndexList;
+    }
+
+    public static String getTableName(String fixtureName) {
+        String cleanedName = fixtureName.replace(":", "_").replace(".", "_").replace("-", "_");
+        return STORAGE_KEY_PREFIX + cleanedName;
+    }
+
+    @Override
+    public String[] getMetaDataFields() {
+        return metaDataFields;
+    }
+
+    @Override
+    public Object getMetaData(String fieldName) {
+        if (fieldName.startsWith(ATTR_PREFIX)) {
+            return root.getAttributeValue(null, getEntryFromCol(fieldName).substring(1));
+        } else if (fieldName.startsWith(ELEM_PREFIX)) {
+            // NOTE PLM: The usage of getChild of '0' below assumes indexes
+            // are only made over entries with multiplicity 0
+            TreeElement child = root.getChild(getEntryFromCol(fieldName), 0);
+            IAnswerData value = child.getValue();
+            if (value == null) {
+                return "";
+            } else {
+                return value.uncast().getString();
+            }
+        }
+        return null;
```

Comment:
It looks like most other implementations of `getMetaData()` throw an `IllegalArgumentException` when they are called with a `fieldName` that doesn't have an indexed column.